### PR TITLE
WIP: Feature/more field types

### LINKED ISSUE #23: Support state with different type of data
Example:

```reason
module FormParams = {
  type state = {
    name: string,
    age: int,
  };
  type fields = [ | `name |  `age ];
  let lens = [
    (`name, s => s.name, (s, name) => {...s, name}),
    (
      `age,
      s => s.age,
      (s, age) => {...s, age}
    ),
  ];
};
```

My current work around is make a variant like this: 

```
type formValue = | Int(int) | String(string); 
```

Is this the right approach? Could reform support this use case?

### PR DESCRIPTION
**Pull request by grsabreu**
TODO 

- [x] Introduce Field states API (Pristine, Dirty, Invalid, Valid, Validating)
- [x] Implement async
- [x] Implement validate-all during onSubmit
- [x] Add support for primitive types

Closes #23 #52 #52

### CODE REVIEW

File: re/Field.re [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  type schema;
+  let lens: list((fields, state => Value.t, (state, Value.t) => state));
+};
+
+module Make = (Config: Config) => {
+  type state =
+    | Validating
+    | Valid
+    | Error(string)
+    | Dirty
```
Comment (grsabreu) [inline re/Field.re]: I'm thinking that `Dirty` is redundant as there is a `Valid` (which means the field changed and was validated)

what you guys think? @Astrocoders/core

Comment (medson10) [reply re/Field.re]: yeah, but we have also the case where the user started to input but the field will be validated just after the submit, in this case the state dirty will be useful. Can be the state here the user started to type, but we don't want valid until he finish

Comment (georgelima) [reply re/Field.re]: I agree with @medson10

**Comment by grsabreu** [comment]
Why this was not merged yet: as all values gotta be a variant now their "scalar" type is not being enforced at compile time.
So still thinking if this will be for the best.

**Comment by grsabreu** [comment]
Thinking in adding GADTs to ensure that back

**Comment by guilhermedecampo** [comment]
super awesome stuff over here. 
Let's record this GADTs change!

**Comment by MarcelCutts** [comment]
This looks great! Can't wait to get my hands on it.

**Comment by grsabreu** [comment]
We're going to publish this as 6.0-beta.x for now for getting community usage feedback

**Comment by shinzui** [comment]
@grsabreu Any change you can release this soon for testing?

**Comment by grsabreu** [comment]
Hey @shinzui, yes! I'm prepare and release it as beta over the weekend
sorry the delay guys

**Comment by grsabreu** [comment]
Thx for the reminding that @OAGr 
We change to a `"namespace": true"` in this branch to avoid name clashing and did not updated the docs yet for that. We'll be pushing this soon

**Comment by grsabreu** [comment]
Closed in favor of #54

**Pull request closed without merging**